src/main/java/rx/observables/AsyncOnSubscribe.java

```diff
@@ -0,0 +1,510 @@
+        S state = generateState();
+        UnicastSubject<Observable<T>> subject = UnicastSubject.<Observable<T>> create();
+        AsyncOuterSubscriber<S, T> outerSubscriberProducer = new AsyncOuterSubscriber<S, T>(this, state, subject);
+        actualSubscriber.add(outerSubscriberProducer);
+        Observable.concat(subject).unsafeSubscribe(Subscribers.wrap(actualSubscriber));
+        actualSubscriber.setProducer(outerSubscriberProducer);
+    }
+
+    private static class AsyncOuterSubscriber<S, T> extends ConcurrentLinkedQueue<Long>implements Producer, Subscription, Observer<Observable<? extends T>> {
+        /** */
+        private static final long serialVersionUID = -7884904861928856832L;
+
+        private volatile int isUnsubscribed;
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<AsyncOuterSubscriber> IS_UNSUBSCRIBED = AtomicIntegerFieldUpdater.newUpdater(AsyncOuterSubscriber.class, "isUnsubscribed");
+
+        private final AsyncOnSubscribe<S, T> parent;
+        private final SerializedObserver<Observable<? extends T>> serializedSubscriber;
+        private final Set<Subscription> subscriptions = new HashSet<Subscription>();
+
+        private boolean hasTerminated = false;
+        private boolean onNextCalled = false;
+
+        private S state;
+
+        private final UnicastSubject<Observable<T>> merger;
+
+        public AsyncOuterSubscriber(AsyncOnSubscribe<S, T> parent, S initialState, UnicastSubject<Observable<T>> merger) {
+            this.parent = parent;
+            this.serializedSubscriber = new SerializedObserver<Observable<? extends T>>(this);
+            this.state = initialState;
+            this.merger = merger;
+        }
+
+        @Override
+        public void unsubscribe() {
+            if (IS_UNSUBSCRIBED.compareAndSet(this, 0, 1)) {
+                // it's safe to process terminal behavior
+                if (isEmpty()) {
+                    parent.onUnsubscribe(state);
+                }
+                for (Subscription s : subscriptions) {
+                    if (!s.isUnsubscribed()) {
+                        s.unsubscribe();
+                    }
+                }
+            }
+        }
+
+        @Override
+        public boolean isUnsubscribed() {
+            return isUnsubscribed != 0;
+        }
+
+        public void nextIteration(long requestCount) {
+            state = parent.next(state, requestCount, serializedSubscriber);
+        }
+
+        @Override
+        public void request(long n) {
+            int size = 0;
+            Long r;
+            synchronized (this) {
+                size = size();
+                add(n);
+                r = n;
+            }
+            if (size == 0) {
+                do {
+                    // check if unsubscribed before doing any work
+                    if (isUnsubscribed()) {
+                        unsubscribe();
```

Comment:
If `isUnsubscribed != 0` then the CAS in `unsubscribe` will fail and nothing happens. If you wanted to delay the unsubscription to the time when nextIteration is not running, then you need a separate indicator, i.e., unsubscribeRequested.